datatypes/src/raster/grid_array.rs

```diff
@@ -0,0 +1,393 @@
+use num_traits::{AsPrimitive, Zero};
+use serde::{Deserialize, Serialize};
+use snafu::ensure;
+
+use crate::error;
+use crate::util::Result;
+
+use super::{
+    BoundedGrid, GridBoundingBox, GridBounds, GridContains, GridIdx, GridIndexAccess,
+    GridIndexAccessMut, GridSize, GridSpaceToLinearSpace,
+};
+
+#[derive(PartialEq, Debug, Copy, Clone, Serialize, Deserialize)]
+pub struct ArrayShape<A>
+where
+    A: AsRef<[usize]>,
+{
+    pub shape_array: A,
+}
+
+impl<A> ArrayShape<A>
+where
+    A: AsRef<[usize]>,
+{
+    pub fn new(shape: A) -> Self {
+        Self { shape_array: shape }
+    }
+
+    pub fn inner_ref(&self) -> &A {
+        &self.shape_array
+    }
+
+    pub fn into_inner(self) -> A {
+        self.shape_array
+    }
+}
+
+pub type ArrayShape1D = ArrayShape<[usize; 1]>;
+pub type ArrayShape2D = ArrayShape<[usize; 2]>;
+pub type ArrayShape3D = ArrayShape<[usize; 3]>;
+
+impl From<[usize; 1]> for ArrayShape1D {
+    fn from(shape: [usize; 1]) -> Self {
+        ArrayShape1D { shape_array: shape }
+    }
+}
+
+impl From<[usize; 2]> for ArrayShape2D {
+    fn from(shape: [usize; 2]) -> Self {
+        ArrayShape2D { shape_array: shape }
+    }
+}
+
+impl From<[usize; 3]> for ArrayShape3D {
+    fn from(shape: [usize; 3]) -> Self {
+        ArrayShape3D { shape_array: shape }
+    }
+}
+
+impl GridSize for ArrayShape1D {
+    type ShapeArray = [usize; 1];
+
+    const NDIM: usize = 1;
+
+    fn axis_size(&self) -> Self::ShapeArray {
+        self.shape_array
+    }
+
+    fn number_of_elements(&self) -> usize {
+        let [a] = self.axis_size();
+        a
+    }
+}
+
+impl GridSpaceToLinearSpace for ArrayShape1D {
+    type IndexArray = [isize; 1];
+
+    fn strides(&self) -> Self::ShapeArray {
+        [1]
+    }
+
+    fn linear_space_index_unchecked<I: Into<GridIdx<Self::IndexArray>>>(&self, index: I) -> usize {
+        let GridIdx([x]) = index.into();
+        let [stride_x] = self.strides();
+        x as usize * stride_x
```

Comment:
make it easier for the compiler by just taking x as index

Reply:
done